casper/src/main/scala/coop/rchain/casper/engine/BlocksRetriever.scala

```diff
@@ -0,0 +1,235 @@
+                if (requestedBlocks.contains(hash)) {
+                  val _ = Log[F].debug(
+                    s"Adding ${PrettyPrinter.buildString(hash)} hash to RequestedBlocks. Already exists."
+                  )
+                  requestedBlocks
+                } else
+                  requestedBlocks +
+                    (hash -> RequestState(
+                      timestamp = now,
+                      peers = if (peerRequested.isEmpty) None.toSet else Set(peerRequested.get),
+                      received = markAsReceived
+                    ))
+              }
+            }
+        )
+
+      override def admitHash(
+          hash: BlockHash,
+          peer: PeerNode,
+          reason: String
+      ): F[Unit] =
+        RequestedBlocks.get(hash).flatMap {
+          case None =>
+            addNewEntry(hash, peerRequested = Some(peer))
+          case Some(_) =>
+            addPeerToWaitingList(
+              hash,
+              peer,
+              reason
+            )
+        } >> RequestedBlocks
+          .get(hash)
+          .flatMap(
+            state =>
+              if (state.get.peers.size == 1)
+                CommUtil[F].requestForBlock(peer, hash) >>
+                  CommUtil[F].broadcastHasBlockRequest(hash)
+              else ().pure[F]
+          )
+
+      override def requestAll(
+          ageThreshold: FiniteDuration
+      ): F[Unit] = {
+
+        def toMap(list: List[(BlockHash, Option[RequestState])]): Map[BlockHash, RequestState] = {
+          val filtered = list.flatMap {
+            case (hash, Some(r)) => List((hash -> r))
+            case (_, None)       => List.empty
+          }
+          Map(filtered: _*)
+        }
+
+        def tryRerequest(
+            hash: BlockHash,
+            requested: RequestState
+        ): F[(BlockHash, Option[RequestState])] =
+          requested.waitingList match {
+            case nextPeer :: waitingListTail =>
+              def modifiedRequested(ts: Long) = requested.copy(
+                timestamp = ts,
+                waitingList = waitingListTail,
+                peers = requested.peers + nextPeer
+              )
+
+              for {
+                _ <- Log[F].debug(
+                      s"Trying ${nextPeer.endpoint.host} to query for ${PrettyPrinter.buildString(hash)} block. " +
+                        s"Remain waiting: ${waitingListTail.map(_.endpoint.host).mkString(", ")}."
+                    )
+                _  <- CommUtil[F].requestForBlock(nextPeer, hash)
+                ts <- Time[F].currentMillis
+              } yield hash -> Option(modifiedRequested(ts))
+            case _ =>
+              val warnMessage = s"Could not retrieve requested block ${PrettyPrinter.buildString(hash)} " +
+                s"from ${requested.peers.mkString(", ")}. Removing the request from the requested blocks list. " +
+                s"Casper will have to re-request the block."
```

Comment:
If node did not receive blocks, e.g. due to networking issue - it will keep asking peers (re-requesting) that reported they have block, one by one. Once block is in Casper, hash is be removed from BlockRetriever scope, so its not subject to this routine.